File: app/src/main/java/nov/me/kanmodel/notes/utils/FileUtils.java
package nov.me.kanmodel.notes.utils;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.text.SimpleDateFormat;

import nov.me.kanmodel.notes.SettingsActivity;

import static android.support.v4.content.FileProvider.getUriForFile;

/**
 * Created by KanModel on 2017/12/26.
 * 文件操作相关函数集
 */

public abstract class FileUtils {
    public static void copy(File src, File dst) throws IOException {
        FileInputStream inStream = new FileInputStream(src);
        FileOutputStream outStream = new FileOutputStream(dst);
        copy(inStream, outStream);
    }

    public static void copy(InputStream inStream, File dst) throws IOException {
        FileOutputStream outStream = new FileOutputStream(dst);
        copy(inStream, outStream);
    }

    public static void copy(InputStream in, OutputStream out) throws IOException {
        int numBytes;
        byte[] buffer = new byte[1024];

        while ((numBytes = in.read(buffer)) != -1)
            out.write(buffer, 0, numBytes);
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public static String saveDatabaseCopy(Context context, File dir)
            throws IOException {
        SimpleDateFormat dateFormat = DateFormats.getBackupDateFormat();
        String date = dateFormat.format(Aid.getNowTime());
        String format = "%s/Note Backup %s.db";
        String filename = String.format(format, dir.getAbsolutePath(), date);

        File db = getDatabaseFile(context);
        File dbCopy = new File(filename);
        FileUtils.copy(db, dbCopy);

        return dbCopy.getAbsolutePath();
    }

    @NonNull
    public static File getDatabaseFile(Context context) {
        String databaseFilename = "Note.db";
        String root = context.getFilesDir().getPath();

        String format = "%s/../databases/%s";
        String filename = String.format(format, root, databaseFilename);

        return new File(filename);
    }

    public static void showSendFileScreen(@NonNull String archiveFilename, SettingsActivity activity) {
        File file = new File(archiveFilename);
        Uri fileUri = getUriForFile(activity, "nov.me.kanmodel.notes", file);

        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_SEND);
        intent.setType("application/zip");
        intent.putExtra(Intent.EXTRA_STREAM, fileUri);
        intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        activity.startActivity(intent);
    }
}

File: app/src/main/java/nov/me/kanmodel/notes/utils/FileUtils.kt
package nov.me.kanmodel.notes.utils

import android.content.Context
import android.content.Intent
import android.support.v4.content.FileProvider.getUriForFile
import nov.me.kanmodel.notes.activity.SettingsActivity
import java.io.*

/**
 * Created by KanModel on 2017/12/26.
 * 文件操作相关函数集
 */
object FileUtils {
    @Throws(IOException::class)
    private fun copy(src: File, dst: File) {
        val inStream = FileInputStream(src)
        val outStream = FileOutputStream(dst)
        copy(inStream, outStream)
    }

    @Throws(IOException::class)
    private fun copy(`in`: InputStream, out: OutputStream) {
        val buffer = ByteArray(1024)
        var numBytes: Int = `in`.read(buffer)

        while (numBytes != -1){
            out.write(buffer, 0, numBytes)
            numBytes = `in`.read(buffer)
        }
    }

    @Throws(IOException::class)
    fun saveDatabaseCopy(context: Context, dir: File): String {
        val filename = String.format("%s/Note Backup %s.db", dir.absolutePath,
                TimeAid.backupDateFormat.format(TimeAid.nowTime))

        val db = getDatabaseFile(context)
        val dbCopy = File(filename)
        copy(db, dbCopy)

        return dbCopy.absolutePath
    }

    private fun getDatabaseFile(context: Context): File {
        val databaseFilename = "Note.db"
        val root = context.filesDir.path
        val filename = String.format("%s/../databases/%s", root, databaseFilename)

        return File(filename)
    }

    fun showSendFileScreen(archiveFilename: String, activity: SettingsActivity) {
        val file = File(archiveFilename)
        val fileUri = getUriForFile(activity, "nov.me.kanmodel.notes", file)

        val intent = Intent()
        intent.action = Intent.ACTION_SEND
        intent.type = "application/zip"
        intent.putExtra(Intent.EXTRA_STREAM, fileUri)
        intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
        activity.startActivity(intent)
    }
}
